lua/dap/ui/variable-tree.lua

```diff
@@ -0,0 +1,246 @@
+--- presentationHint?: VariablePresentationHint
+--- evaluateName?: string
+--- variablesReference: number
+--- namedVariables?: number
+--- indexedVariables?: number
+--- memoryReference?: string
+function M.write_variables(buf, variables, line, column, win)
+  line = line or 0
+  column = column or 0
+
+  local state = variable_buffers[buf]
+  local win_width = api.nvim_win_get_width(win)
+  local win_height = api.nvim_win_get_height(win)
+  local indent = string.rep(" ", column)
+
+  local sorted_variables = {}
+  for _, v in pairs(variables) do
+    table.insert(sorted_variables, v)
+  end
+  table.sort(
+    sorted_variables,
+    function(a, b)
+      return a.name < b.name
+    end
+  )
+  local max_textlength = 0
+
+  for _, v in pairs(sorted_variables) do
+    state.line_to_variable[line] = v
+    local text =
+      indent ..
+      v.name ..
+        (non_empty_sequence(v.value) and M.variable_value_separator .. v.value or "") ..
+          " " .. (v.type and (" " .. v.type) or "")
+
+    local splitted_text = vim.split(text, "\n")
+    api.nvim_buf_set_lines(buf, line, line + #splitted_text, false, splitted_text)
+
+    if M.multiline_variable_display then
+      line = line + #splitted_text
+    else
+      api.nvim_buf_set_lines(buf, line, line + 1, false, {table.concat(splitted_text, " ")})
+      line = line + 1
+    end
+
+    if v.variables and variable_buffers[buf].expanded[v.variablesReference] then
+      local inner_max
+      line, inner_max = M.write_variables(buf, v.variables, line, column + 2, win)
+      max_textlength = math.max(max_textlength, inner_max)
+    end
+    max_textlength = math.max(max_textlength, #text)
+  end
+  if win then
+    if win_width < max_textlength then
+      api.nvim_win_set_width(win, math.min(max_textlength + 2, M.max_win_width))
+    end
+    if win_height < line then
+      api.nvim_win_set_height(win, line)
+    end
+  end
+  api.nvim_buf_set_lines(buf, line, -1, false, {})
+  return line, max_textlength
+end
+
+function M.hover(session)
+  if not session then
+    return
+  end
+  if not session.stopped_thread_id then
+    return
+  end
+
+  local scopes = utils.get_at_path(session, "current_frame.scopes")
+
+  if vim.tbl_contains(vim.api.nvim_list_wins(), floating_win) then
+    vim.api.nvim_set_current_win(floating_win)
+  else
+    local cword = vim.fn.expand("<cword>")
+    local variable
+    for _, s in pairs(scopes) do
+      variable = s.variables and s.variables[cword]
+      if variable then
+        break
+      end
+    end
+    if variable then
+      floating_buf, floating_win = vim.lsp.util.open_floating_preview({""}, "dap-variable-ui")
+      create_variable_buffer(floating_buf, floating_win, session, {variable})
+    else
+      print('"' .. cword .. '" not found!')
+    end
+  end
+end
+
+function M.open_sidebar(session, steal_focus, winopts, variables)
+  if not session then
+    return
+  end
+  if not session.stopped_thread_id then
+    return
+  end
```

Comment:
I'm not sure about having the session as argument here.

I'd expect to be able to keep the side bar open across debug sessions and that it would get updated on a new session.

Could the sidebar register a `custom_event_handlers` for `event_initialized` to get access to the session and unsubscribe if the buffer gets detached?

Reply:
Yes, I planned this as a future extension. The buffer could subscribe to certain events when it should update variables. I will have a look into how to do this.

I could subscribe to `event_subscribe` to create a global sidebar that always shows the current scopes. How it's written right now, a window is opened to inspect a certain set of variables. So you can have a window for certain scopes of different frames, or for a certain set variables that you want to watch.

I could write some functions actually using the functionality and subscribing to session protocols to implement a local scopes or a watch window.